src/GafferScene/BranchCreator.cpp

```diff
@@ -170,38 +228,137 @@ class BranchCreator::BranchesData : public IECore::Data
 			}
 		}
 
-		const IECore::PathMatcher &destinationPaths() const
+		bool empty() const
 		{
-			return m_destinationPaths;
+			return m_root->children.empty() && !m_root->sourcePaths;
 		}
 
-		// Maps from destination path to a list of source paths.
-		const map<ScenePlug::ScenePath, vector<ScenePlug::ScenePath>> &sourcePaths() const
+		const Location *locationOrAncestor( const ScenePlug::ScenePath &path ) const
 		{
-			return m_sourcePaths;
+			const Location *result = m_root.get();
+			for( const auto &name : path )
+			{
+				const auto it = result->children.find( name );
+				if( it != result->children.end() )
+				{
+					result = it->second.get();
+				}
+				else
+				{
+					break;
+				}
+			}
+			return result;
+		}
+
+		const Location::SourcePaths &sourcePaths( const ScenePlug::ScenePath &destination ) const
+		{
+			const Location *location = locationOrAncestor( destination );
+			assert( location->depth == destination.size() );
+			if( !location->sourcePaths )
+			{
+				throw IECore::Exception( boost::str( boost::format( "No source paths found for destination \"%1%\"" ) % ScenePlug::pathToString( destination ) ) );
+			}
+			return *location->sourcePaths;
+		}
+
+		template<typename F>
+		void visitDestinations( F &&f ) const
+		{
+			visitLocationsWalk(
+				[ &f ] ( const ScenePlug::ScenePath &path, const Location *location ) {
+					if( location->sourcePaths )
+					{
+						f( path, *location->sourcePaths );
+					}
+				},
+				ScenePath(),
+				m_root.get()
+			);
 		}
 
 	private :
 
+		template<typename F>
+		void visitLocationsWalk( F &&f, const ScenePlug::ScenePath &path, Location *location ) const
+		{
+			f( path, location );
+
+			ScenePlug::ScenePath childPath = path; childPath.push_back( InternedString() );
+			for( const auto &child : location->children )
+			{
+				childPath.back() = child.first;
+				visitLocationsWalk( f, childPath, child.second.get() );
+			}
+		}
+
 		static void hashBranch( const BranchCreator *branchCreator, const ScenePlug::ScenePath &path, IECore::MurmurHash &h )
 		{
-			branchCreator->destinationPlug()->hash( h );
+			h.append( path.data(), path.size() );
+			h.append( path.size() );
+
+			ScenePlug::ScenePath destination = ScenePlug::stringToPath( branchCreator->destinationPlug()->getValue() );
+			h.append( destination.data(), destination.size() );
+			h.append( destination.size() );
+
+			const ScenePlug::ScenePath existing = closestExistingPath( branchCreator->inPlug(), destination );
+			h.append( existing.data(), existing.size() );
+			h.append( existing.size() );
```

Comment:
Sorry, I hadn't thought about this previously - but a couple questions about this:

* why switch to using stringToPath on the destination?  Was there a problem with it?  It certainly feels faster?
* thinking through why we need to append size ... oh the problem would be if the end of one path happened to match the beginning of the other.  Fair enough.

Reply:
We need `destination` as a path anyway, to use with `closestExistingPath`, so it seemed reasonable to hash it that way too.

Reply:
Yeah, that's fair.  I guess it could be a win in the long run if we do stored hashes for InternedStrings at some point.